### ISSUE TITLE
TypeScript allowing usage of Array member that should not be available based on target.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [microsoft/TypeScript#33355]
In `@pulumi/pulumi` we have our tsconfig set to have `"target": "es6"`: https://github.com/pulumi/pulumi/blob/d3475c02b82dfbbeed1bae737929075cf2d73d93/sdk/nodejs/tsconfig.json#L4

Based on hte docs here https://www.typescriptlang.org/docs/handbook/compiler-options.html, that's equivalent to `For --target ES6: DOM,ES6,DOM.Iterable,ScriptHost`.

However, despite this, when we compile, we find that some non-es6 features seem to be allowed by TS.  For example, i can write:

```ts
const v = ["a"];
v.includes("a");
```

Which compiles without error.  There is also no issue shown in vscode.  This is strange since vscode shows `.includes` as coming from `lib.es2016.array.include.d.ts`.   Why would an es2016 property be available in an es6 target/lib?

We're def very confused by this and could use some advice on what we're missing here.

**Comment by CyrusNajmabadi** [comment]
Tagging @DanielRosenwasser

**Comment by RyanCavanaugh** [comment]
Matt Ellis pinged me about this offline. You should sync with him - the root cause is a `lib` import coming from a `.d.ts` file

**Comment by CyrusNajmabadi** [comment]
He is at lunch!   :D

In the meantime, i looked through the compiler options.  I couldn't find anything to help really trace down how this got pulled in.  Want to teach me how to fish @RyanCavanaugh ?  I did try `--traceResolution`.  But that didn't seem to produce any entries for `lib.es2016` or `Array` like i was hoping.

**Comment by AnyhowStep** [comment]
It might be coming from one of your dependencies.

It's a problem I sometimes have, too.

I'll set my `lib` to something like `es5` but then one of my dependencies has a triple-slash reference (or whatever it's called) to `es2015` and it pollutes my project.

**Comment by RyanCavanaugh** [comment]
There isn't something to help with non-module resolution.

`node.d.ts` has (correctly) a `/// <reference lib="es2016" />` directive (or equivalent), because the version of node it represents includes all the ES2016 APIs and needs to reference some of the types therein. So your project claims (indirectly) to be running on a version of node that has those APIs present.

The right end-game here is that someone needs to fix `node.d.ts` to not greedily pull in more `lib` than it actually needs

**Comment by AnyhowStep** [comment]
I'm pretty sure I've seen **dev dependencies** pull `es20xx` references into my project, even though I'm targeting `es5`.

It's possible that I want to use `es5` for the library but use `es20xx` for tests.
